packages/grid/x-data-grid-pro/src/hooks/features/infiniteLoader/useGridInfiniteLoader.ts

```diff
@@ -22,62 +20,72 @@ export const useGridInfiniteLoader = (
   >,
 ): void => {
   const visibleColumns = useGridSelector(apiRef, gridVisibleColumnDefinitionsSelector);
   const currentPage = useGridVisibleRows(apiRef, props);
-  const rowsMeta = useGridSelector(apiRef, gridRowsMetaSelector);
-  const contentHeight = Math.max(rowsMeta.currentPageTotalHeight, 1);
+  const observer = React.useRef<IntersectionObserver>();
+  const previousY = React.useRef(0);
+  const previousRatio = React.useRef(0);
+  const previousNode = React.useRef<HTMLElement>();
 
-  const isInScrollBottomArea = React.useRef<boolean>(false);
+  const handleLoadMoreRows = React.useCallback(
+    ([entry]: IntersectionObserverEntry[]) => {
+      const currentY = entry.boundingClientRect.y;
+      const currentRatio = entry.intersectionRatio;
+      const isIntersecting = entry.isIntersecting;
 
-  const handleRowsScrollEnd = React.useCallback(
-    (scrollPosition: GridScrollParams) => {
-      const dimensions = apiRef.current.getRootDimensions();
+      // Scrolling down check
+      if (currentY < previousY.current || isIntersecting) {
+        if (currentRatio >= previousRatio.current) {
+          const viewportPageSize = apiRef.current.getViewportPageSize();
+          const rowScrollEndParam: GridRowScrollEndParams = {
+            visibleColumns,
+            viewportPageSize,
+            visibleRowsCount: currentPage.rows.length,
+          };
+          apiRef.current.publishEvent('rowsScrollEnd', rowScrollEndParam);
+        }
+      }
+
+      previousY.current = currentY;
+      previousRatio.current = currentRatio;
+    },
+    [apiRef, visibleColumns, currentPage.rows.length],
+  );
 
+  const lastVisibleRowRef = React.useCallback<GridInfiniteLoaderApi['unstable_lastVisibleRowRef']>(
+    (node) => {
       // Prevent the infite loading working in combination with lazy loading
-      if (!dimensions || props.rowsLoadingMode !== 'client') {
+      if (props.rowsLoadingMode !== 'client') {
         return;
       }
 
-      const scrollPositionBottom = scrollPosition.top + dimensions.viewportOuterSize.height;
-      const viewportPageSize = apiRef.current.getViewportPageSize();
-
-      if (scrollPositionBottom < contentHeight - props.scrollEndThreshold) {
-        isInScrollBottomArea.current = false;
+      if (observer.current) {
+        observer.current.disconnect();
       }
 
-      if (
-        scrollPositionBottom >= contentHeight - props.scrollEndThreshold &&
-        !isInScrollBottomArea.current
-      ) {
-        const rowScrollEndParam: GridRowScrollEndParams = {
-          visibleColumns,
-          viewportPageSize,
-          visibleRowsCount: currentPage.rows.length,
-        };
-        apiRef.current.publishEvent('rowsScrollEnd', rowScrollEndParam);
-        isInScrollBottomArea.current = true;
+      observer.current = new IntersectionObserver(handleLoadMoreRows, {
+        threshold: props.lastVisibleRowThreshold,
+      });
+
+      if (node && node !== previousNode.current) {
```

Comment:
It looks like the original issue is now reproducible again https://github.com/mui/mui-x/issues/4371
See this demo: https://codesandbox.io/s/infiniteloadinggrid-material-demo-forked-3nf8wf

Reply:
Would be great to have a unit test for this issue specifically

Reply:
Actually, it should work like this. If you have fewer rows than what the viewport can hold the callback is invoked. If this doesn't happen, the callback will never be called because there is no ability to scroll.

Reply:
> Actually, it should work like this. If you have fewer rows than what the viewport can hold the callback is invoked.

Yes, the issue is that you're stuck with 6 rows unable to load more rows

Reply:
yup this calls for a test indeed.